docs/reusing-typechecking-results.md

```diff
@@ -0,0 +1,217 @@
+- It's not clear if it's going to be for better or for worse, since the signature generation penalty can be bigger than the benefit of having them.
+- Even if it's for better, such a small improvement won't be worse the hassle.
+
+Therefore, I think it's not the way to go here.
+
+<details>
+
+<summary>Numbers, for completeness</summary>
+
+```
+Normal compilation:
+1. Typecheck: 23.7531
+2. Typecheck: 27.4234
+3. Typecheck: 24.5202
+
+Normal compilation + generating signatures (added extra `ReportTime` for measuring the latter):
+1. Typecheck: 26.5991, Siggen: 1.9369
+2. Typecheck: 25.0246, Siggen: 1.7517
+3. Typecheck: 27.0057, Siggen: 1.9093
+
+Compilation with given signatures:
+1. Typecheck: 25.7741
+2. Typecheck: 26.2904
+3. Typecheck: 24.3852
+
+Compilation with given 80% signatures + generating 20% signatures:
+1. Typecheck: 24.1338, Siggen: 0.5284
+2. Typecheck: 27.6037, Siggen: 0.7526
+3. Typecheck: 25.6967, Siggen: 0.6244
+```
+
+</details>
+
+
+## Implementation plan
+
+The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages where each stage brings gains in some scenarios.
+
+**Stage 1 - force-gen and compare typechecking graphs**
+
+We already create the TC graph in `main1` and we are able to dump it (in Mermaid) via the compiler flags.
+
+That means we can force-gen and save the graph which will allow us to skip retypechecking if we detect that the graph is not changed. We should also track all the compilation information (the argument string) and the last update times of the files.
+
+This step won't bring big observable benefits, yet it will create necessary MSBuild hooks to communicate the intermediate files folder towards the compiler, add time-based and hash-based cache invalidation logic, and create testing rails which will include the `clean` and `rebuild` tests to make sure the cache is easily invalidated on demand.
+
+**Stage 2 - skip retypechecking for some files**
+
+In `main1`, we do unpickling (= deserializing) and pickling (= serializing) of the code. This currently happens on the module/namespace basis. 
+
+In this stage, we will fully implement pickling and unpickling of all typechecked files (`CheckedImplFile` per file as well as all other outputs of `main1` which cannot be cheaply recreated like `topAttrs` and `CcuThunk`). Parts of the pickling/unpickling can reuse the primitives already existing in `TypedTreePickle` and built upon them. So we can save acquired typechecking information to the intermediate folder after the typechecking - and for the files not needing retypechecking as per graph diff, skip the phase, instead restoring the typechecking info.
```

Comment:
This should probably mention there is some compiler global state that gets modified during a type check. So probably we will also need some sort of recording and replaying of side-effects on top of the pickled result.

Reply:
That's a good point. Added there, thanks!